### PR TITLE
Add customizability for impact report generation

### PR DESCRIPTION
**Pull request by lucernae**
### Add callback option for impact report preproces

In some cases, we might want to have a pre process exactly before the report generated. E.g, change QGIS composition context, change settings, etc.

Also handle QLR layer input for Impact Report.

### Checklist:

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Add to the changelog in metadata.txt if it's a new feature
- [x] Unit test for new code added
- [x] Request someone to review or test your PR

### CODE REVIEW

**Comment by lucernae** [comment]
Whoops, I didn't know we have different impact function implementation.
Need to fix that too.

**Comment by lucernae** [comment]
@Gustry I mean, we have different impact function instance: regular one, and for multi exposure. So I had to provide the same method signature for both :D.

I tried to search by inheritance previously, but it seems MultiExposureImpactFunction does not inherit from ImpactFunction. Was it intended?

**Comment by lucernae** [comment]
@ismailsunni @myarjunar @Gustry Travis can someone tell me how to spot the flake8 errors mentioned by travis? Hahaha... I can't find it on travis and it doesn't show up on my local machine :(.

**Comment by lucernae** [comment]
I think the test failed because now the default lines changed from 5 to 6. Here's the excerpt of the test output in my local machine:

```
Test Flake8 result:

-----------
Flake8 issues
-----------
3.6.0 (mccabe: 0.6.1, pycodestyle: 2.4.0, pyflakes: 2.0.0) CPython 2.7.12 on Linux

End of test Flake8 result.
```

I will create a commit to check this first.

File: safe/gis/test/test_tools.py [hunk 1]
```diff
@@ -59,6 +59,19 @@ def test_load_layer_from_uri(self):
...
+        # This QLR contains a file based reference layer to
+        # small_grid.geojson, and it should work the same way
+        # as if we load small_grid.geojson
+        # In practice we could put the non file-based layer (PostGIS/WFS)
+        # and load it from QLR
+        path = standard_data_path(
+            'gisv4', 'aggregation', 'small_grid.qlr')
+        layer, purpose = load_layer(path)
+        self.assertTrue(layer.isValid())
+        self.assertEqual(layer.name(), 'small_grid')
```
[the remark this answers is not in the document]
Comment (lucernae) [reply safe/gis/test/test_tools.py]: eeeh, not a chance :p

**Comment by lucernae** [comment]
Huh, I got this:

```
Test if the code is Flake8 compliant. ... make: *** No rule to make target 'flake8'.  Stop.
Test Flake8 result:

End of test Flake8 result.
FAIL
```

No make target???

**Comment by lucernae** [comment]
Ok. Found it. 
`TestPythonPep` mostly uses relative path to find Makefile. But the relative path uses os.getcwd (the current working dir) and it may change during the test (somehow) depending on which previous test files were inspected.
So, I'm going to refactor these things to use relative path to find Makefile, from the current file location instead.

**Comment by lucernae** [comment]
@Gustry @ismailsunni @myarjunar If you have any objection/suggestion for the flake8 fix, I will do it in the next PR. I'm going to merge this one first.

**Comment by ismailsunni** [comment]
Thanks @lucernae 
Can you make a PR or ticket to apply it to develop branch also?

**Pull request merged**